README.md
# vscode-speech-to-text README

This is the README for your extension "vscode-speech-to-text". After writing up a brief description, we recommend including the following sections.

## Features

Describe specific features of your extension including screenshots of your extension in action. Image paths are relative to this README file.

For example if there is an image subfolder under your extension project workspace:

\!\[feature X\]\(images/feature-x.png\)

> Tip: Many popular extensions utilize animations. This is an excellent way to show off your extension! We recommend short, focused animations that are easy to follow.

## Usage
1. `cmd/ctrl+shift+p` to open the command palette.
2. Run the `Speech to Text: Dictate` command. This will start a web server at `localhost:9000` and a WebSocket server at `localhost:9001`.
3. `localhost:9000` will automatically open in your default browser. If it doesn't, open it manually in Chrome or another browser that [supports the Web Speech API](https://developer.mozilla.org/en-US/docs/Web/API/SpeechRecognition#Browser_compatibility). This will connect to the WebSocket server and also ask for permission to use your device's microphone. You need to allow this to start using speech to text. Keep this page open (Note: keeping several tabs with this page open simultaneously could cause malfunctioning) and proceed to the next step.
4. (WIP) Speak into your microphone. Each phrase you speak will be compared against a predefined list of recognized phrases. If there are any matches, they will be listed inline in your editor.

## Requirements

If you have any requirements or dependencies, add a section describing those and how to install and configure them.

- Internet connection. The Web Speech API requires this. [See the note under Speech recognition](https://developer.mozilla.org/en-US/docs/Web/API/Web_Speech_API/Using_the_Web_Speech_API). The speed of the extension depends on the speed of your connection.

## Extension Settings

Include if your extension adds any VS Code settings through the `contributes.configuration` extension point.

For example:

This extension contributes the following settings:

* `myExtension.enable`: enable/disable this extension
* `myExtension.thing`: set to `blah` to do something

## Known Issues

Calling out known issues can help limit users opening duplicate issues against your extension.

## Release Notes

Users appreciate release notes as you update your extension.

## Resources
- [https://developer.mozilla.org/en-US/docs/Web/API/Web_Speech_API/Using_the_Web_Speech_API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Speech_API/Using_the_Web_Speech_API)
- [https://code.visualstudio.com/api/get-started/your-first-extension](https://code.visualstudio.com/api/get-started/your-first-extension)
- [https://code.visualstudio.com/api/references/vscode-api](https://code.visualstudio.com/api/references/vscode-api)
- [https://code.visualstudio.com/api/references/vscode-api#WorkspaceEdit](https://code.visualstudio.com/api/references/vscode-api#WorkspaceEdit)
- [https://stackoverflow.com/questions/39569993/vs-code-extension-get-full-path](https://stackoverflow.com/questions/39569993/vs-code-extension-get-full-path)
- [https://github.com/microsoft/vscode/issues/111](https://github.com/microsoft/vscode/issues/111) (Extension API: cursor position (determining current location and moving it) #111)
- [https://github.com/Microsoft/vscode/issues/35000](https://github.com/Microsoft/vscode/issues/35000) (Trigger a Completion Item from key binding in vscode extensions #35000)
- [https://github.com/TalAter/annyang](https://github.com/TalAter/annyang) (Not currently using, but might be worth a look?)

### 1.0.0

Initial release of ...

### 1.0.1

Fixed issue #.

### 1.1.0

Added features X, Y, and Z.

-----------------------------------------------------------------------------------------------------------

## Working with Markdown

**Note:** You can author your README using Visual Studio Code.  Here are some useful editor keyboard shortcuts:

* Split the editor (`Cmd+\` on macOS or `Ctrl+\` on Windows and Linux)
* Toggle preview (`Shift+CMD+V` on macOS or `Shift+Ctrl+V` on Windows and Linux)
* Press `Ctrl+Space` (Windows, Linux) or `Cmd+Space` (macOS) to see a list of Markdown snippets

### For more information

* [Visual Studio Code's Markdown Support](http://code.visualstudio.com/docs/languages/markdown)
* [Markdown Syntax Reference](https://help.github.com/articles/markdown-basics/)

**Enjoy!**


CHANGELOG.md
# Change Log

All notable changes to the "vscode-speech-to-text" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

client/index.js
const socket = new WebSocket('ws://localhost:9001');
var SpeechRecognition = (SpeechRecognition || webkitSpeechRecognition);
var SpeechGrammarList = (SpeechGrammarList || webkitSpeechGrammarList);
var SpeechRecognitionEvent = (SpeechRecognitionEvent || webkitSpeechRecognitionEvent);
const recognition = new SpeechRecognition();
const speechGrammarList = new SpeechGrammarList();
const css = [ // Only required for words that aren't recognized by the default speech recognition service.
	// 'background-color',
	// 'color',
	// 'display',
	// 'font-family',
	// 'font-size',
	// 'indent',
	// 'list-style',
	// 'opacity', 
	// 'resize'
	'eenie meenie'
];
const grammar = '#JSGF V1.0; grammar colors; public <color> = ' + css.join(' | ') + ' ;';
const outputEl = document.getElementById('output');

speechGrammarList.addFromString(grammar, 1);
recognition.continuous = true; // Keep recording results.
recognition.grammars = speechGrammarList;
recognition.interimResults = false; // Return non-final results. Fast, but inaccurate.
recognition.lang = 'en-US'; // 'nb-NO' for norwegian.
recognition.maxAlternatives = 1;

recognition.addEventListener('result', e => {
	const result = e.results[e.resultIndex][0].transcript.trim();
	socket.send(result);
	outputEl.innerHTML += `<li>${result}</li>`;
});
recognition.addEventListener('end', e => {
	/**
	 * Restart speech recognition when it stops, 
	 * as it seems to stop after X time without speech.
	 * https://stackoverflow.com/questions/42895760/how-to-stop-and-restart-web-speech-api-correctly
	 * https://stackoverflow.com/questions/34818154/webkitspeechrecognition-stops-recording-randomly
	 * https://stackoverflow.com/questions/38702797/webkitspeechrecognition-stops-doesnt-fire-onend-at-random
	 */
	recognition.start();
});

// todo: remove this when done testing:
['audioend', 'audiostart', 'end', 'error', 'nomatch', 'result', 'soundend', 'soundstart', 'speechend', 'speechstart', 'start']
.forEach(name => recognition.addEventListener(name, e => console.log(e)));

socket.addEventListener('open', e => {
	recognition.start();
	// todo: remove this when done testing:
	// socket.send('for');
});

// todo: remove this when done testing:
// setTimeout(() => {
// 	location.reload();
// }, 2000);

src/extension.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';
const path = require('path');
const WebSocket = require('ws');
const express = require('express');
const open = require('open');
const app = express();
const port = 9000;
const socketPort = 9001;

app.use('/', express.static(path.join(__dirname, '../client')));
app.listen(port, () => {
	vscode.window.showInformationMessage(`[Speech to Text] Server running at localhost:${port}`);
	open(`http://localhost:${port}`);
});

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {
	// Use the console to output diagnostic information (console.log) and errors (console.error)
	// This line of code will only be executed once when your extension is activated
	console.log('[Speech to Text] Congratulations, your extension "vscode-speech-to-text" is now active!');

	// The command has been defined in the package.json file
	// Now provide the implementation of the command with registerCommand
	// The commandId parameter must match the command field in package.json
	let disposable = vscode.commands.registerCommand('stt.dictate', () => {
		// The code you place here will be executed every time your command is executed
		const wss = new WebSocket.Server({port: socketPort});

		wss.on('connection', (socket: any) => {
			// Display a message box to the user
			console.log('[Speech to Text] New WebSocket connection');
			vscode.window.showInformationMessage('[Speech to Text] New WebSocket connection');

			socket.on('message', (data: string) => {
				console.log(`[Speech to Text] New WebSocket message: ${data as string}`);

				/* vscode.languages.registerCompletionItemProvider('javascript', {
					provideCompletionItems: (doc: vscode.TextDocument, pos: vscode.Position, token: vscode.CancellationToken, context: vscode.CompletionContext): vscode.ProviderResult<vscode.CompletionList> => {
						console.log('[Speech to Text] inside');
						return new vscode.CompletionList([
							{
								label: 'I am a method',
								kind: 1
							},
							{
								label: 'I am a function',
								kind: 2
							},
							{
								label: 'I am a property',
								kind: 9
							},
							{
								label: 'I don\'t have a kind',
							},
						]);
					}
				}, '.'); */ // todo: trigger this programmatically. Update: don' think we need this at all since we can use the code below instead?

				/**
				 * Add text to currently open document/file.
				 * todo: Add support for special characters (enter, space, tab, backspace, arrows, etc). Use vscode.commands.
				 * todo: Add support for removing text. Use vscode.commands.
				 * todo: Add support for multiple selections? Use vscode.commands.
				 */
				const path = vscode.window.activeTextEditor?.document.fileName;
				const cursorPos = vscode.window.activeTextEditor?.selection.active;
				if (path) {
					const edit = new vscode.WorkspaceEdit();
					const uri = vscode.Uri.file(path);
					const position = new vscode.Position(cursorPos?.line||0, cursorPos?.character||0);
					edit.insert(uri, position, data);
					vscode.workspace.applyEdit(edit).then(res => vscode.commands.executeCommand('editor.action.triggerSuggest'));
					// vscode.commands.getCommands(false).then(commands => console.log(JSON.stringify(commands))); // Get all available commands.
				}
			});
		});
	});

	context.subscriptions.push(disposable);
}

// this method is called when your extension is deactivated
export function deactivate() {}
